src/MIDebugEngine/Engine.Impl/DebugUnixChildProcess.cs

```diff
@@ -0,0 +1,260 @@
+                    _vforkBp = null;
+                }
+                _state = State.Init;
+            }
+        }
+
+        public async Task SetBreakAtMain()
+        {
+            Results results = await _process.MICommandFactory.BreakInsert("main", null);
+            var bkpt = results.Find("bkpt");
+            if (bkpt is ValueListValue)
+            {
+                _mainBreak = ((ValueListValue)bkpt).Content[0].FindString("number");
+            }
+            else
+            {
+                _mainBreak = bkpt.TryFindString("number");
+            }
+        }
+
+        public async Task<bool> Stopped(Results results)
+        {
+            string reason = null;
+
+            switch (_state)
+            {
+                case State.Enabled:
+                    reason = results.TryFindString("reason");
+                    if (reason == "fork")
+                    {
+                        int threadId = results.FindInt("thread-id");
+                        _newpid = results.FindInt("newpid");
+                        _state = State.AtFork;
+                        await _process.Step(threadId, VisualStudio.Debugger.Interop.enum_STEPKIND.STEP_OUT, VisualStudio.Debugger.Interop.enum_STEPUNIT.STEP_LINE);
+                        break;
+                    }
+                    else if (reason == "vfork")
+                    {
+                        int threadId = results.FindInt("thread-id");
+                        _newpid = results.FindInt("newpid");
+                        await _process.MICommandFactory.Set("schedule-multiple", "on");
+                        await _process.MICommandFactory.Catch("exec", onlyOnce: true);
+                        var thread = await _process.ThreadCache.GetThread(threadId);
+                        _state = State.AtVfork;
+                        await _process.Continue(thread);
+                        break;
+                    }
+                    return false;
+                case State.AtFork:
+                    await ProcessChild();
+                    break;
+                case State.AtVfork:
+                    if ("exec" == results.TryFindString("reason"))
+                    {
+                        await _process.MICommandFactory.Set("schedule-multiple", "off");
+                        await RunChildToMain();
+                    }
+                    else
+                    {
+                        // sometimes gdb misses the breakpoint at exec and execution will proceed to a breakpoint in the child
+                        _process.Logger.WriteLine("Missed catching the exec after vfork. Spawning the child's debugger.");
+                        _state = State.AtExec;
+                        goto missedExec;
+                    }
+                    break;
+                case State.AtSignal:    // both child and parent are stopped
+                    return await DetachAndContinue();
+                case State.AtExec:
```

Comment:
Is that because we have left the parent suspended? Or because vfork should suspend the parent until the child calls an exec function? If the later, I wasn't clear on if Linux would allow other threads in the process to run or not.